src/core/weightedGraph.test.js

```diff
@@ -0,0 +1,56 @@
+// @flow
+
+import * as Weights from "./weights";
+import {Graph} from "./graph";
+import * as WeightedGraph from "./weightedGraph";
+import * as GraphTest from "./graphTestUtil";
+
+describe("core/weightedGraph", () => {
+  function expectEqual(wg1, wg2) {
+    expect(wg1.graph.equals(wg2.graph)).toBe(true);
+    expect(wg1.weights).toEqual(wg2.weights);
+  }
+  it("empty produces an empty WeightedGraph", () => {
+    const {weights, graph} = WeightedGraph.empty();
+    expect(graph.equals(new Graph())).toBe(true);
+    expect(weights).toEqual(Weights.empty());
+  });
+  describe("toJSON/fromJSON", () => {
```

Comment:
Depending on the nested `Compatible<T>` question above.
Would appreciate some test coverage for the top-level and underlying versions we expect to support.
Similar to the Project tests
https://github.com/sourcecred/sourcecred/blob/325487a3ea569e1a0d176b60f39245fb2243804d/src/core/project.test.js#L46

Reply:
Are you saying I should add a test case that validates that it upgrades from 0.1.0 (ie. the current version)? 

Personally, I don't think this is needed. Graph (and WeightedGraph) is re-generated ephemerally, but not persisted, so we don't actually need upgrade handling yet. (For project, which is actually config and not data, it's reasonable that we invest more effort in handling upgrades.)

Reply:
Not the upgrade path per say. But `Compatible<T>` adds flow coverage for the shape of the body. And having the compat info header. Not which version that compat info should have.

Testing we have composed the correct versions seems useful for this nesting case.

Reply:
I'm sorry, I still don't understand what you're asking for. Would you be willing to add this as a follow-on PR? This is PR is upstream of a lot of work I want to do in a timely fashion, so I'm feeling some impatience to not block this on non-structural issues.

Reply:
What I mean is, there is no test coverage for the contents of COMPAT_INFO. Neither through flow nor unit tests. Since we're composing them I think it would make sense to test we composed them correctly.

```diff
[
- COMPAT_INFO,
  {
    graphJSON: [
-     COMPAT_INFO,
      graph
    ],
    weightsJSON: [
-     COMPAT_INFO,
      weights
    ]
  }
]
```

Or to put it differently, I would like to see a test *fail* when we accidentally make an incompatible version upgrade one one of the parts.

```diff
Expected:
[
  {type: "sourcecred/weightedGraph", version: "0.1.0"},
+ {type: "sourcecred/graph", version: "0.8.0"},
  {type: "sourcecred/weights", version: "0.2.0"},
]

Actual:
[
  {type: "sourcecred/weightedGraph", version: "0.1.0"},
- {type: "sourcecred/graph", version: "0.9.0"},
  {type: "sourcecred/weights", version: "0.2.0"},
]
```

Reply:
Sorry, I still don't understand your ask well enough to faithfully add this test. Are you suggesting that we have a snapshot test that verifies that the composed compats were as expected? Or an upgrade test that validates that we can correctly load a compat with the current version?

Reply:
I'm going to merge this, but please feel free to send a PR adding this. I'm genuinely curious to see what you had in mind, and suspect it will be a valuable improvement.